tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,467 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#include <linux/kernel.h>
+#include <linux/module.h>
+
+#include "tempesta_fw.h"
+#include "apm.h"
+#include "log.h"
+#include "server.h"
+
+MODULE_AUTHOR(TFW_AUTHOR);
+MODULE_DESCRIPTION("Tempesta round-robin scheduler");
+MODULE_VERSION("0.3.1");
+MODULE_LICENSE("GPL");
+
+#define	TFW_RATIO_TIMEOUT	(HZ/20)		/* The timer periodicity. */
+
+/*
+ * The size of pool of TfwRatioRated{} entries for RCU substitution.
+ * It's best when it's one more than the real number of CPU cores.
+ */
+#define TFW_RATIO_EPOOLSZ	(16 + 1)
+
+/*
+ * The size of scale that is used to range back-end servers.
+ * It's most effective in calculations when it's a power of two. 
+ */
+#define TFW_RATIO_SCALE_64	(1 << 6)
+#if TFW_SG_MAX_SRV > TFW_RATIO_SCALE_64
+#define TFW_RATIO_SCALE		TFW_SG_MAX_SRV
+#else
+#define TFW_RATIO_SCALE		TFW_RATIO_SCALE_64
+#endif
+
+/**
+ * List of connections to an upstream server.
+ * Connections can go up and down during failover process. Only
+ * fully established connections are considered by the scheduler.
+ */
+typedef struct {
+	atomic64_t		counter;
+	size_t			conn_n;
+	TfwServer		*srv;
+	TfwConnection 		*conns[TFW_SRV_MAX_CONN];
+} TfwRatioSrv;
+
+/*
+ * RCU-fortified array of pointers to TfwRatioSrv{} ranged by each
+ * back-end server's weight. The number of times a server is found
+ * in the array is defined by the server's weight relative to other
+ * back-end servers in the group.
+ */
+typedef struct {
+	struct rcu_head		rcu;
+	size_t			listsz;
+	TfwRatioSrv		*list[TFW_RATIO_SCALE];
+	atomic_t		free;
+} TfwRatioRated;
+
+/**
+ * List of upstream servers.
+ * The list is considered static, i.e. all servers, either dead
+ * or alive, are present in the list during the whole run-time.
+ * That may change in the future.
+ */
+typedef struct {
+	atomic64_t		counter;
+	size_t			srv_n;
+	atomic_t		rearm;
+	struct timer_list	timer;
+	TfwRatioSrv		srvs[TFW_SG_MAX_SRV];
+	TfwRatioRated		epool[TFW_RATIO_EPOOLSZ];
+	TfwRatioRated __rcu	*rated;
+} TfwRatioSrvList;
+
+static void
+tfw_sched_ratio_calc(int *weight, int *ratio, size_t sz)
+{
+	size_t s;
+	unsigned long calc[sz], sum = 0, unit;
+
+	for (s = 0; s < sz; ++s) {
+		calc[s] = ((1 << 16) * TFW_RATIO_SCALE) / weight[s];
+		sum += calc[s];
+	}
+	unit = ((1 << 16) * TFW_RATIO_SCALE) / sum;
+	for (s = 0; s < sz; ++s)
+		ratio[s] = (calc[s] * unit) / (1 << 16) ? : 1;
+}
+
+static inline void
+tfw_sched_fill_rated(TfwSrvGroup *sg, int *ratio, TfwRatioRated *rated)
+{
+	size_t s, i, n;
+	TfwRatioSrvList *sl = sg->sched_data;
+
+	for (s = 0, i = 0; s < sl->srv_n; ++s)
+		for (n = 0; n < ratio[s]; ++n, ++i)
+			rated->list[i] = &sl->srvs[s];
```

Comment:
`TfwRatioSrv->list` has size 64 (1 << 6), which is just 2 times larger than `TFW_SG_MAX_SRV`, so it seems the array can be overrun on large ratios.